tests/unittests/HyphenTest.cpp

```diff
@@ -0,0 +1,322 @@
+  EXPECT_EQ(hyphens, (vector<bool>{0, 1, 0, 0, 0, 1, 0, 0}));
+}
+
+/// Map a lower-case letter to an input index in the range 0-26.
+/// Use 0 to represent any characters outside the a-z range.
+static size_t mapLetter(char l) {
+  unsigned d = l - unsigned('a');
+  return d < 26 ? d + 1 : 0;
+}
+
+TEST(HyphenTest, mapLetter) {
+  EXPECT_EQ(mapLetter('a'), 1);
+  EXPECT_EQ(mapLetter('d'), 4);
+  EXPECT_EQ(mapLetter('z'), 26);
+  EXPECT_EQ(mapLetter('.'), 0);
+}
+
+/// Map a 6-letter window of a word to an input tensor using a one-hot encoding.
+///
+/// The tensor must be N x 6 x 27: batch x position x letter.
+static void mapLetterWindow(const string &window, size_t idx,
+                            Handle<float> tensor) {
+  EXPECT_EQ(window.size(), 6);
+  for (size_t row = 0; row < 6; row++) {
+    size_t col = mapLetter(window[row]);
+    tensor.at({idx, row, col}) = 1;
+  }
+}
+
+// Training data consisting of pre-hyphenated common words.
+const vector<const char *> TrainingData{
+    "ad-mi-ni-stra-tion",
+    "ad-mit",
+    "al-low",
+    "al-though",
+    "an-i-mal",
+    "any-one",
+    "ar-rive",
+    "art",
+    "at-tor-ney",
+    "be-cause",
+    "be-fore",
+    "be-ha-vior",
+    "can-cer",
+    "cer-tain-ly",
+    "con-gress",
+    "coun-try",
+    "cul-tural",
+    "cul-ture",
+    "de-cide",
+    "de-fense",
+    "de-gree",
+    "de-sign",
+    "de-spite",
+    "de-velop",
+    "di-rec-tion",
+    "di-rec-tor",
+    "dis-cus-sion",
+    "eco-nomy",
+    "elec-tion",
+    "en-vi-ron-men-tal",
+    "es-tab-lish",
+    "ev-ery-one",
+    "ex-actly",
+    "ex-ec-u-tive",
+    "ex-ist",
+    "ex-pe-ri-ence",
+    "ex-plain",
+    "fi-nally",
+    "for-get",
+    "hun-dred",
+    "in-crease",
+    "in-di-vi-du-al",
+    "it-self",
+    "lan-guage",
+    "le-gal",
+    "lit-tle",
+    "lo-cal",
+    "ma-jo-ri-ty",
+    "ma-te-rial",
+    "may-be",
+    "me-di-cal",
+    "meet-ing",
+    "men-tion",
+    "mid-dle",
+    "na-tion",
+    "na-tional",
+    "oc-cur",
+    "of-fi-cer",
+    "par-tic-u-lar-ly",
+    "pat-tern",
+    "pe-riod",
+    "phy-si-cal",
+    "po-si-tion",
+    "pol-icy",
+    "pos-si-ble",
+    "pre-vent",
+    "pres-sure",
+    "pro-per-ty",
+    "pur-pose",
+    "re-cog-nize",
+    "re-gion",
+    "re-la-tion-ship",
+    "re-main",
+    "re-sponse",
+    "re-sult",
+    "rea-son",
+    "sea-son",
+    "sex-ual",
+    "si-mi-lar",
+    "sig-ni-fi-cant",
+    "sim-ple",
+    "sud-den-ly",
+    "sum-mer",
+    "thou-sand",
+    "to-day",
+    "train-ing",
+    "treat-ment",
+    "va-ri-ous",
+    "value",
+    "vi-o-lence",
+};
+
+TEST(HyphenTest, network) {
+  // Convert the training data to word windows and labels.
+  vector<string> words;
+  vector<bool> hyphens;
+  for (auto *hword : TrainingData) {
+    dehyphenate(hword, words, hyphens);
+  }
+
+  // This depends on the training data, of course.
+  const size_t numSamples = 566;
+  ASSERT_EQ(hyphens.size(), numSamples);
```

Comment:
The following code is the random-shuffle loop `for (size_t i = 0; i != numSamples; i++)`, so it would be unsafe to continue if the two vectors don't have the required size. I think `ASSERT_EQ` is right here since it aborts immediately on failure.

Reply:
EXPECT_EQ - should also fail the test immediately. Up to you, just have not found a lot of usages of asserts/ASSERT_EQ where regular gtest EXPECTs could be used.

Reply:
EXPECT_EQ will mark the test as failing, but it keeps running. This means you can have multiple failures from a single test run, which can be useful. See [the googletest docs explanation](https://github.com/google/googletest/blob/master/googletest/docs/Primer.md#assertions).